src/Rule/ForbidNotNormalizedTypeRule.php

```diff
@@ -0,0 +1,564 @@
+     */
+    private function processMultiTypePhpParserNode(
+        ComplexType $multiTypeNode,
+        Scope $scope,
+        string $identification
+    ): array
+    {
+        $innerTypeNodes = array_values($multiTypeNode->types);
+        $multiTypeNodeString = $this->printPhpParserNode($multiTypeNode);
+
+        $errors = [];
+        $countOfNodeTypes = count($innerTypeNodes);
+
+        foreach ($innerTypeNodes as $i => $iValue) {
+            for ($j = $i + 1; $j < $countOfNodeTypes; $j++) {
+                $typeNodeA = $iValue;
+                $typeNodeB = $innerTypeNodes[$j];
+
+                $typeA = $scope->getFunctionType($typeNodeA, false, false);
+                $typeB = $scope->getFunctionType($typeNodeB, false, false);
+
+                $typeNodeAString = $this->printPhpParserNode($typeNodeA);
+                $typeNodeBString = $this->printPhpParserNode($typeNodeB);
+
+                if ($typeA->isSuperTypeOf($typeB)->yes()) {
+                    $errors[] = RuleErrorBuilder::message("Found non-normalized type {$multiTypeNodeString} for {$identification}: {$typeNodeBString} is a subtype of {$typeNodeAString}.")
+                        ->line($multiTypeNode->getLine())
+                        ->build();
+                    continue;
+                }
+
+                if ($typeB->isSuperTypeOf($typeA)->yes()) {
+                    $errors[] = RuleErrorBuilder::message("Found non-normalized type {$multiTypeNodeString} for {$identification}: {$typeNodeAString} is a subtype of {$typeNodeBString}.")
+                        ->line($multiTypeNode->getLine())
+                        ->build();
+                }
+            }
+        }
+
+        return $errors;
+    }
+
+    private function printPhpParserNode(PhpParserNode $node): string
+    {
+        $nodeCopy = clone $node;
+        $nodeCopy->setAttribute('comments', []); // avoid printing with surrounding line comments
+        return $this->phpParserPrinter->prettyPrint([$nodeCopy]);
+    }
+
+    /**
+     * @param UnionTypeNode|IntersectionTypeNode $multiTypeNode
+     * @return list<RuleError>
+     */
+    private function processMultiTypePhpDocNode(
+        TypeNode $multiTypeNode,
+        NameScope $nameSpace,
+        ?string $identification,
+        int $line
+    ): array
+    {
+        $errors = [];
+        $innerTypeNodes = array_values($multiTypeNode->types); // ensure list
+        $forWhat = $identification !== null ? " for $identification" : '';
+
+        if ($multiTypeNode instanceof IntersectionTypeNode) {
+            foreach ($multiTypeNode->types as $type) {
+                if ($type instanceof UnionTypeNode) {
+                    $dnf = $this->typeNodeResolver->resolve($multiTypeNode, $nameSpace)->describe(VerbosityLevel::typeOnly());
+
+                    $errors[] = RuleErrorBuilder::message("Found non-normalized type {$multiTypeNode}{$forWhat}: this is not disjunctive normal form, use {$dnf}")
```

Comment:
Forcing DNF feels a bit out-of-scope. DNF type can be a lot longer, so it's not always better.

Reply:
I think this part can be configurable.